lib/clickhousex/protocol.ex

```diff
@@ -74,84 +68,88 @@ defmodule Clickhousex.Protocol do
     end
   end
 
-  @doc false
-  @spec reconnect(new_opts :: Keyword.t(), state) :: {:ok, state}
-  def reconnect(new_opts, state) do
-    with :ok <- disconnect("Reconnecting", state),
-         do: connect(new_opts)
+  # TODO find a cleaner way
+  @spec ensure_connected(state) :: {:ok, state} | {:error, Exception.t()}
+  defp ensure_connected(state) do
+    %{conn: conn, conn_opts: conn_opts} = state
+
+    if Client.open?(conn) do
+      {:ok, state}
+    else
+      case connect(conn_opts) do
+        {:ok, _state} = ok -> ok
+        {:error, _conn, reason} -> {:error, reason}
+      end
+    end
```

Comment:
I'm reconnection right in the callbacks to avoid returning `{:disconnect, error, state}` from them since that would result in an error being logged, which we are trying to avoid.

Reply:
hmm yeah this seems a bit hacky and working against what DBConnection is doing. Like I said would be good to get some feedback from that project for why DBConnection works this way and whether we're on the right track.